services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/TopicCleanupService.java

```diff
@@ -401,4 +446,40 @@ public static List<PubSubTopic> extractVersionTopicsToCleanup(
         })
         .collect(Collectors.toList());
   }
+
+  private List<PubSubTopic> collectDanglingTopics(Map<PubSubTopic, Long> pubSubTopicsRetentions) {
+    List<PubSubTopic> topicsTobeCleanup = new ArrayList<>();
+    Set<PubSubTopic> kafkaTopics = apacheKafkaAdminAdapter.get().listAllTopics();
+    for (Map.Entry<PubSubTopic, Long> entry: pubSubTopicsRetentions.entrySet()) {
+      PubSubTopic pubSubTopic = entry.getKey();
+      if (!kafkaTopics.contains(pubSubTopic)) {
+        try {
+          String storeName = pubSubTopic.getStoreName();
+          String clusterDiscovered = admin.discoverCluster(storeName).getFirst();
+          Store store = admin.getStore(clusterDiscovered, storeName);
+          LOGGER.info("Find topic discrepancy case: {}", pubSubTopic);
+          if (pubSubTopic.isRealTime()) {
+            if (!store.isHybrid()) {
+              LOGGER.info("Will remove real-time dangling topic {}", pubSubTopic);
+              topicsTobeCleanup.add(pubSubTopic);
+            }
+          } else if (pubSubTopic.isVersionTopicOrStreamReprocessingTopic() || pubSubTopic.isViewTopic()) {
+            int versionNum = Version.parseVersionFromKafkaTopicName(pubSubTopic.getName());
+            if (!store.containsVersion(versionNum)) {
+              LOGGER.info("Will remove dangling version topic {}", pubSubTopic);
+              topicsTobeCleanup.add(pubSubTopic);
+            }
```

Comment:
We submitted a change recently to preserve VT data when DIV fatal issue happens. Can we double check that it won't be considered as dandling topics thus got deleted here?

Reply:
Here dangling topic deletion only focus on the topics gap inside pub sub system component, as long as we did not try to delete the existing topics, so that topic will not be evaluated by this feature. Moreover, this feature only focus on the dangling topics are not valid based on the criteria(not vaild version, not valid real-time topic ....) anymore.